### PR TITLE
Add masquerade as specific student support

### PR DESCRIPTION
**Pull request by stvstnfrd**
This must merge _before_ the platform PR.

Screenshot_2020-07-13_03-19-44

Screenshot_2020-07-13_03-20-35

Screenshot_2020-07-13_03-20-54

Screenshot_2020-07-13_03-21-24

Screenshot_2020-07-13_03-21-51

Screenshot_2020-07-13_03-22-36

Screenshot_2020-07-13_03-22-56

Screenshot_2020-07-13_03-29-32

Screenshot_2020-07-13_03-29-19

### CODE REVIEW

File: src/instructor-toolbar/masquerade-widget/data/api.js [hunk 1]
```diff
@@ -7,8 +7,8 @@ export async function getMasqueradeOptions(courseId) {
   return camelCaseObject(data);
 }
 
-export async function postMasqueradeOptions(courseId, data) {
+export async function postMasqueradeOptions(courseId, payload) {
   const url = new URL(`${getConfig().LMS_BASE_URL}/courses/${courseId}/masquerade`);
-  const { response } = await getAuthenticatedHttpClient().post(url.href, data);
-  return camelCaseObject(response);
+  const { data } = await getAuthenticatedHttpClient().post(url.href, payload);
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/data/api.js:12]: We have to unpack `data`, not `response`, else we get nothing.

This was "broken" before, but we weren't using the response yet, so it was irrelevant.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx [hunk 2]
```diff
@@ -4,40 +4,49 @@ import React, {
...
-      && selected.groupId === (this.props.groupId || null)
-      && selected.role === this.props.role
-    );
+    /* eslint-disable arrow-body-style */
+    const isEqual = [
+      'groupId',
+      'role',
+      'userName',
+      'userPartitionId',
+    ].reduce((accumulator, currentValue) => {
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx:49]: Is there a more idiomatic way of doing this w/ JS/React?

This is essentially an equality check of the two objects, looking only at select properties.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx [hunk 3]
```diff
@@ -4,40 +4,49 @@ import React, {
...
+      return accumulator && (
+        this.props[currentValue] === this.props.selected[currentValue]
+      );
+    }, true);
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx:53]: Initialize the state to `true`.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx [hunk 4]
```diff
@@ -4,40 +4,49 @@ import React, {
...
-      && selected.role === this.props.role
-    );
+    /* eslint-disable arrow-body-style */
+    const isEqual = [
+      'groupId',
+      'role',
+      'userName',
+      'userPartitionId',
+    ].reduce((accumulator, currentValue) => {
+      return accumulator && (
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx:50]: Start short-circuiting once any check returns `false`.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx [hunk 5]
```diff
@@ -4,40 +4,49 @@ import React, {
...
       global.location.reload();
     });
+    return true;
   }
 
   isSelected() {
-    const selected = this.props.selected || {};
-    const isEqual = (
-      selected.userPartitionId === (this.props.userPartitionId || null)
-      && selected.groupId === (this.props.groupId || null)
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx:38]: These two coalescing operators are unnecessary;
types are consistent for our purposes.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx [hunk 6]
```diff
@@ -4,40 +4,49 @@ import React, {
...
-      && selected.role === this.props.role
-    );
+    /* eslint-disable arrow-body-style */
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx:43]: Intentionally ignoring this...

I get why it's complaining (the arrow function really is single-line), but this just feels too long for a single line;
I think readability would suffer.

Open to be overruled though.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx [hunk 7]
```diff
@@ -4,40 +4,49 @@ import React, {
...
     const {
-      courseId,
       groupId,
       role,
       userName,
       userPartitionId,
+      userNameInput,
     } = this.props;
     const payload = {};
+    // TODO: Close the dropdown
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx]: I went through Paragon docs, but couldn't quite tell if this was directly supported or not.

The dropdown _does_ automatically close if you click outside of the element, but not when the `<input />` grabs `focus()`.

We should close the dropdown when the text input is displayed, but I hadn't got that working.

This wasn't an issue w/ the other masquerade options, since we reload the page, instead of showing an input.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx [hunk 8]
```diff
@@ -4,40 +4,49 @@ import React, {
...
+    // return true;
     if (role) {
       payload.role = role;
     }
     if (groupId) {
       payload.group_id = parseInt(groupId, 10);
       payload.user_partition_id = parseInt(userPartitionId, 10);
     }
-    if (userName) {
-      payload.user_name = userName;
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx:27]: We don't need to include the `userName` in the payload, because we'll make specific-user masquerade requests separately (when they submit the text input, not click the dropdown option).

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx [hunk 9]
```diff
@@ -58,16 +66,17 @@ class MasqueradeWidgetOption extends Component {
...
         onClick={(event) => this.handleClick(event)}
+        data-toggle="dropdown"
       >
         {groupName}
       </Dropdown.Item>
     );
   }
 }
 MasqueradeWidgetOption.propTypes = {
-  courseId: PropTypes.string.isRequired,
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx:68]: We don't currently need this.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx [hunk 10]
```diff
@@ -4,40 +4,48 @@ import React, {
 import PropTypes from 'prop-types';
 import { Dropdown } from '@edx/paragon';
 
-import { postMasqueradeOptions } from './data/api';
-
 class MasqueradeWidgetOption extends Component {
   handleClick() {
     const {
-      courseId,
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx:12]: We weren't using this any longer.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx [hunk 11]
```diff
@@ -55,14 +95,25 @@ class MasqueradeWidget extends Component {
...
+          </Dropdown.Button>
+          <Dropdown.Menu>
+            {options}
+          </Dropdown.Menu>
+        </Dropdown>
+        <MasqueradeUserNameInput
+          className="flex-shrink-0"
+          onError={() => this.onError.apply(this, arguments)}
+          onSubmit={(payload) => this.onSubmit(payload)}
+          ref={(input) => { this.userNameInput = input; }}
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx]: We create a `ref` to it, so we can call its `toggle` method from in the `MasqueradeWidgetOption`s.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx [hunk 12]
```diff
@@ -55,14 +95,25 @@ class MasqueradeWidget extends Component {
...
-        <Dropdown.Button>
-          View this course as
-        </Dropdown.Button>
-        <Dropdown.Menu>
-          {options}
-        </Dropdown.Menu>
-      </Dropdown>
+      <>
+        <Dropdown
+          className="flex-shrink-1"
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx]: I'll be totally honest, I haven't fully groked `flex` yet, so some of this was trial-and-error.
Let me know if it's off.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx [hunk 13]
```diff
@@ -57,17 +65,18 @@ class MasqueradeWidgetOption extends Component {
       <Dropdown.Item
         className={className}
         href="#"
-        onClick={(event) => this.handleClick(event)}
+        onClick={(event) => this.onClick(event)}
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx:73]: Just renaming for consistency.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx [hunk 14]
```diff
@@ -4,40 +4,48 @@ import React, {
...
     }
     if (groupId) {
       payload.group_id = parseInt(groupId, 10);
       payload.user_partition_id = parseInt(userPartitionId, 10);
     }
-    if (userName) {
-      payload.user_name = userName;
-    }
-    postMasqueradeOptions(courseId, payload).then(() => {
+    this.props.onSubmit(payload).then(() => {
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx:36]: Delegate the actual submission to the helper provided to us by the parent.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx [hunk 15]
```diff
@@ -4,40 +4,48 @@ import React, {
 import PropTypes from 'prop-types';
 import { Dropdown } from '@edx/paragon';
 
-import { postMasqueradeOptions } from './data/api';
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx:7]: Let the parent handle all of the AJAX calls.

File: src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+    });
+  }
+
+  render() {
+    return (this.state.shouldShowUserNameInput
+      && (
+        <>
+          &nbsp;
+          <Input
+            autoFocus
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx]: Magically grab focus when displayed, yay!

File: src/alerts/masquerade-error-alert/hooks.js [hunk 17]
```diff
@@ -0,0 +1,20 @@
+/* eslint-disable import/prefer-default-export */
+import { useMemo } from 'react';
+import { useModel } from '../../generic/model-store';
+import { useAlert } from '../../generic/user-messages';
+
+export function useMasqueradeErrorAlert(courseId) {
+  const masquerade = useModel('masquerades', courseId);
```
Comment (stvstnfrd) [inline src/alerts/masquerade-error-alert/hooks.js]: Note: this data-store model does not exist. I wasn't able to get this all hooked up yet, so I wasn't able to use the built-in alert lists framework. IMO, the alerts look better than the current implementation (a message in the toolbar, similar to existing experience).

If we hook up the data-store, this _should_ just start working. If we don't do that though, let's remove these alerts.

File: src/alerts/masquerade-status-alert/MasqueradeStatusAlert.jsx [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+  ALERT_TYPES,
+} from '../../generic/user-messages';
+
+function MasqueradeStatusAlert({ payload }) {
+  const {
+    message,
+  } = payload;
+  const isVisible = !!message;
+  return isVisible && (
+    <Alert type={ALERT_TYPES.INFO}>
```
Comment (stvstnfrd) [inline src/alerts/masquerade-status-alert/MasqueradeStatusAlert.jsx]: It's unfortunate we need to spin up all of this machinery, just to say, "display a one-off alert message with arbitrary text"...

File: src/alerts/masquerade-status-alert/hooks.js [hunk 19]
```diff
@@ -0,0 +1,20 @@
+/* eslint-disable import/prefer-default-export */
+import { useMemo } from 'react';
+import { useModel } from '../../generic/model-store';
+import { useAlert } from '../../generic/user-messages';
+
+export function useMasqueradeStatusAlert(courseId) {
+  const masquerade = useModel('masquerades', courseId);
```
Comment (stvstnfrd) [inline src/alerts/masquerade-status-alert/hooks.js]: Per above, this doesn't exist.
The alert should be remove if unused.

File: src/instructor-toolbar/InstructorToolbar.jsx [hunk 20]
```diff
@@ -51,23 +51,21 @@ export default function InstructorToolbar(props) {
...
           </div>
         )}
-        &nbsp;
         {urlStudio && (
-          <div className="flex-shrink-0">
+          <div className="flex-shrink-0 mx-1 my-1">
             <a className="btn d-block btn-outline-light" href={urlStudio}>View in Studio</a>
           </div>
         )}
-        &nbsp;
```
Comment (stvstnfrd) [inline src/instructor-toolbar/InstructorToolbar.jsx:68]: I learned my lesson, got rid of the spaces, and used proper margins :)

File: src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+    const payload = {
+      role: 'student',
+      user_name: event.target.value,
+    };
+    this.props.onSubmit(payload).then((data) => {
+      if (data && data.success) {
+        global.location.reload();
+      } else {
+        const error = (data && data.error) || '';
+        this.onError(error);
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx:32]: Proxy errors back up to the parent component, via the wrapper/prop function.

File: src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+      }
+    }).catch((error) => {
+      this.onError(error);
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx]: TODO: We probably don't want this displayed verbatim; should probably be replaced with a generic message.

File: src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+    return true;
+  }
+
+  render() {
+    return (
+      <Input
+        autoFocus
+        className="flex-shrink-1"
+        defaultValue=""
+        label="TODO"
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx]: TODO: whoops...

File: src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+        onKeyPress={(event) => this.onKeyPress(event)}
+        placeholder="username or email"
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx]: TODO: i18n?

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx [hunk 25]
```diff
@@ -4,47 +4,111 @@ import React, {
...
       options: [],
+      error: null,
+      shouldShowUserNameInput: false,
+      status: null,
     };
   }
 
   componentDidMount() {
     getMasqueradeOptions(this.courseId).then((data) => {
+      // There are two failure states:
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx]: TODO: Remove comment block prior to merge.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx [hunk 26]
```diff
@@ -4,47 +4,111 @@ import React, {
...
+      // There's not much we can do to recover;
+      // if we can't fetch masquerade status,
+      // assume it's disabled/unavailable.
       console.error('Unable to get masquerade options', response);
     });
   }
 
-  parseAvailableOptions(payload) {
-    const data = payload || {};
+  onError(message) {
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx:50]: If we were to hook into the alert system, this is where we'd trigger the error.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx [hunk 27]
```diff
@@ -4,47 +4,111 @@ import React, {
...
+
+  async onSubmit(payload) {
+    this.setState({
+      shouldShowUserNameInput: false,
+    });
+    const options = await postMasqueradeOptions(this.courseId, payload);
+    return options;
+  }
+
+  onSuccess(data) {
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx:67]: If we were to hook into the alert system, this is where we'd trigger the status.

**Review by stvstnfrd — COMMENTED** [review]
In the process of typing up hand-off notes, I was able to work out a few final kinks, so mostly everything left _should_ be superficial/optional.

TL;DR: Dropdown doesn't always autoclose and the alert/error/status styling could use some color/spacing tweaking.

Outstanding/unaddressed issues:
- Selecting an option from the dropdown does not close the dropdown.
  - This seems to be the default behavior from Paragon and it is how the existing experience works (in the MFE), specifically, that the dropdown _does_ close if you click outside the component, but not if you select an item from _within_ the dropdown. It doesn't even close when it blurs.
  - It wasn't an issue previously, since we refresh the entire page immediately (for the previously existing options). But now, it looks weird if the dropdown doesn't close and you select "view as a specific student".
  - Hopefully this is something easy enough (adding/removing a class, etc.), as it'll improve the UX.
- I couldn't get the messages hooked into the alerting framework.
  - This is probably just me not groking the data-store properly to configure/use it, but I couldn't get it working.
  - The relevant alerting components exist, they just don't do anything w/out the data-store backing. If we can't hook them up, we should revert that commit (and remove the alerting components entirely)
  - So I implemented them as messages directly w/in the instructor toolbar.
    - I'm using the bootstrap alert classes, but the default styling is a bit jarring; it's too light. This could use some tweaking.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx [hunk 28]
```diff
@@ -4,47 +4,111 @@ import React, {
...
-        const options = this.parseAvailableOptions(data);
-        this.setState({
-          options,
-        });
+        this.onSuccess(data);
       } else {
-        console.warn('Unable to get masquerade options', data);
+        // This was explicitly denied by the backend;
+        // assume it's disabled/unavailable.
+        console.error('Unable to get masquerade options', data);
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx]: Nit: I didn't _really_ mean to change this `s/warn/error`...

**Comment by stvstnfrd** [comment]
Screenshot_2020-07-13_03-42-45

Screenshot_2020-07-13_03-49-47

**Comment by stvstnfrd** [comment]
CC: @marcotuts

File: src/alerts/masquerade-error-alert/MasqueradeErrorAlert.jsx [hunk 29]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import {
+  Alert,
+  ALERT_TYPES,
+} from '../../generic/user-messages';
+
+function MasqueradeErrorAlert({ payload }) {
```
Comment (davidjoy) [inline src/alerts/masquerade-error-alert/MasqueradeErrorAlert.jsx]: From our conversation: Because this error type is basically a generic error and isn't 'special' or custom in any way, we can do away with this error component and just use the default one.

File: src/alerts/masquerade-status-alert/MasqueradeStatusAlert.jsx [hunk 30]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import {
+  Alert,
+  ALERT_TYPES,
+} from '../../generic/user-messages';
+
+function MasqueradeStatusAlert({ payload }) {
```
Comment (davidjoy) [inline src/alerts/masquerade-status-alert/MasqueradeStatusAlert.jsx]: Same goes for this one.

File: src/alerts/masquerade-status-alert/MasqueradeStatusAlert.jsx (hunk 18 above)
Comment (davidjoy) [reply src/alerts/masquerade-status-alert/MasqueradeStatusAlert.jsx]: Agreed! :)

File: src/alerts/masquerade-error-alert/hooks.js (hunk 17 above)
Comment (davidjoy) [reply src/alerts/masquerade-error-alert/hooks.js]: As we talked about, we can luckily dispense with this hook completely.  If we use the UserMessagesContext in MasqueradeWidget, we can use the `add` function it exposes directly at the time when we know we want to add the message.  

Background on this hook/class component code like it: A hook like this is primarily useful when we need to react to a change in properties to inform a message being added/removed, which is an ugly/difficult thing to do in React.  In a class component, we would use the `componentDidUpdate` lifecycle method and compare the current props with the previous props to understand if something changed, and then conditionally create an alert based on it.  In either case, it's annoying.  Luckily we don't have need of that here. :)

File: src/courseware/course/Course.jsx [hunk 31]
```diff
@@ -5,6 +5,8 @@ import { getConfig } from '@edx/frontend-platform';
```
Comment (davidjoy) [inline src/courseware/course/Course.jsx]: All the changes in this file can go away too, if we drop the above custom alerts!

File: src/instructor-toolbar/InstructorToolbar.jsx [hunk 32]
```diff
@@ -51,23 +51,21 @@ export default function InstructorToolbar(props) {
   return (
     <div className="bg-primary text-light">
       <div className="container-fluid py-3 d-md-flex justify-content-end align-items-center">
-        <div className="flex-grow-1">
+        <div className="flex-grow-1 d-md-flex mx-1 my-1">
```
Comment (davidjoy) [inline src/instructor-toolbar/InstructorToolbar.jsx]: FWIW - I noticed when we were demoing it that the input box seemed to be aligned a bit higher than the other elements, probably because it's height is different.  Since these are all in a flexbox, you should be able to add `align-items-center` as a class here and it _should_ improve it.  I think.

File: src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx [hunk 33]
```diff
@@ -0,0 +1,56 @@
+import React, {
+  Component,
+} from 'react';
+import PropTypes from 'prop-types';
+import { Input } from '@edx/paragon';
+
+class MasqueradeUserNameInput extends Component {
+  constructor(props) {
+    super(props);
+    this.onError = this.props.onError;
```
Comment (davidjoy) [inline src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx]: Saving this as an instance variable is a bit non-standard, fwiw.  I think in general you'd just call it from `this.props`.

Another thing people do if they want to reuse something in props a lot is to destructure it in the function they're in, i.e.:

```
const { onError } = this.props; // at the top of onSubmit, perhaps
```

File: src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx [hunk 34]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  onKeyPress(event) {
```
Comment (davidjoy) [inline src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx:15]: This is a naming nit, but in terms of method/prop naming scheme... my general preference and that shown in a lot of the react documentation is for handler methods to be named with a `handle` prefix, i.e., `handleKeyPress`.  

This is less documented in the community and feels a bit more subjective, but when passing a handler down to a sub component (like `onError` above), I usually name those something like `errorHandler` since it's a reference to the handler.  It also lets me know at a glance that it's coming from the outside of the component, as opposed to one defined in it.

And finally, doing the above makes it so that the "on" prefix only really gets used in built-in HTML mouse/keyboard attributes on the JSX itself.  

I kinda like the division of having those three separate schemes.  Makes it easy to tell what's coming from where.

File: src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx [hunk 35]
```diff
@@ -0,0 +1,56 @@
...
+    if (event.key === 'Enter') {
+      return this.onSubmit(event);
+    }
+    return true;
+  }
+
+  onSubmit(event) {
```
Comment (davidjoy) [inline src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx:22]: Separate from the above, if you declare these handler methods like this:

```
onSubmit = (event) => {
}
```

You can avoid any potential binding issues without having to go put a `this.onSubmit = this.onSubmit.bind(this);` in the constructor.

It's only a real _problem_ if you're passing them to child components, but I try to either do the constructor binding or declare them as arrow functions to avoid it coming back to bite me later when it starts getting passed around as the code gets more complex. :)

File: src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx (hunk 21 above)
Comment (davidjoy) [reply src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx:32]: I haven't thought through the ownership implications, but if the parent component handles the response - rather than this input component - then we don't have to worry about doing any of the proxying.  Arguably this component could just be "dumber" and let the parent be "smarter" w/r/t handling requests and such.

It's also mildly interesting that we're waiting for a promise to resolve on one of our props... I don't think I've done that before.  I don't see any reason why it wouldn't work, technically, fwiw. :)

File: src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx (hunk 24 above)
Comment (davidjoy) [reply src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx]: Yeah, prolly. 

```
import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';
import messages from './messages.js';

....

this.props.intl.formatMessage(messages['masquerade.input.placeholder'])

...

MasqueradeUserNameInput.propTypes = {
  ...
  intl: intlShape,
}

export default injectIntl(MasqueradeUserNameInput);
```

Or thereabouts. :)

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx (hunk 28 above)
Comment (davidjoy) [reply src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx]: Wonder if we should log it externally if this happens? 🤔

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx (hunk 26 above)
Comment (davidjoy) [reply src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx:50]: Agreed!  If we add UserMessagesContext as a `contextType` on this component, we can then do `this.context.add()` here and give it a type of error, our message (or a code), etc.

File: src/instructor-toolbar/masquerade-widget/data/api.js (hunk 1 above)
Comment (davidjoy) [reply src/instructor-toolbar/masquerade-widget/data/api.js:12]: Yeah, this looks right now. :)

File: src/instructor-toolbar/InstructorToolbar.jsx [hunk 36]
```diff
@@ -51,23 +51,21 @@ export default function InstructorToolbar(props) {
   return (
     <div className="bg-primary text-light">
       <div className="container-fluid py-3 d-md-flex justify-content-end align-items-center">
-        <div className="flex-grow-1">
+        <div className="align-items-center flex-grow-1 d-md-flex mx-1 my-1">
```
Comment (stvstnfrd) [inline src/instructor-toolbar/InstructorToolbar.jsx:54]: @davidjoy `align-items-center` added.

File: src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx (hunk 33 above)
Comment (stvstnfrd) [reply src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx]: How "non-standard" are we talking here?! ;)
mostly jk, but curious...

Hrmm, my thinking was that I wanted the caller(s) to be implementation agnostic;
this allows an `onError` method to be defined directly on this class itself, if it wanted to handle the error with_out_ passing it back up (eg: `console.log` and swallow the error).

so we can always just call `this.onError` and never worry about where it's defined.

It's the architectural equivalent of saying "pretty please use my handler, but I respect that you could use your own";
here, we're playing along ;)

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx (hunk 27 above)
Comment (stvstnfrd) [reply src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx:67]: Done.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx (hunk 26 above)
Comment (stvstnfrd) [reply src/instructor-toolbar/masquerade-widget/MasqueradeWidget.jsx:50]: Done.

File: src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx (hunk 34 above)
Comment (stvstnfrd) [reply src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx:15]: That is good to know; I appreciate consistency.

> And finally, doing the above makes it so that the "on" prefix only really gets used in built-in HTML mouse/keyboard attributes on the JSX itself.

Just to clarify that unless I'm misunderstanding, this here _is_ a built-in HTML attribute;
we're binding an `onKeyPress` event to the input element.

File: src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx (hunk 35 above)
Comment (stvstnfrd) [reply src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx:22]: Re: above: I consider this to also be a native(ish) DOM event handler, as it takes a DOM `event` object as its `argument`. If we were writing semantic HTML (don't get me started today ;) ), it'd be a proper `FORM.onSubmit` :)

**Review by stvstnfrd — COMMENTED** [review]
@davidjoy i18n and other feedback addressed.

The one thing I haven't got working yet is the dropdown hiding.
Should I just make this a native dropdown/select until the bootrap/paragon update lands?

Bummer, I liked this UI!

File: src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx (hunk 34 above)
Comment (davidjoy) [reply src/instructor-toolbar/masquerade-widget/MasqueradeUserNameInput.jsx:15]: I meant that in the sense that the JSX would read:

```onKeyPress={handleKeyPress}```

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx (hunk 2 above)
Comment (davidjoy) [reply src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx:49]: I think this fits your use case?  I'd try using https://lodash.com/docs/4.17.15#isEqual or something like it.

Comment (davidjoy) [reply src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx:49]: You could first filter the objects by the properties you want, maybe?  

`reduce` is confusing at the best of times, IMHO. ;)

**Review by davidjoy — APPROVED** [review]

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx (hunk 2 above)
Comment (stvstnfrd) [reply src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx:49]: I took a look and yeah, between needing to filter the props anyway, plus the library calls, I think we'd have a final result that's at least as long/complicated as this one. And since this is natively supported _AND_ the function does one thing and is (I think!) well-named, I'm going to leave this alone for now.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx [hunk 37]
```diff
@@ -4,40 +4,47 @@ import React, {
...
-
 class MasqueradeWidgetOption extends Component {
-  handleClick() {
+  onClick(event) {
+    // TODO: Remove this hack when we upgrade to Paragon's Dropdown
+    event.target.parentNode.parentNode.click();
```
Comment (stvstnfrd) [inline src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx:16]: @davidjoy I'm not particularly proud of this, but it did save me from converting the component ;)
and since this functionality has been added to the newest Paragon release, we don't need to worry about swapping the UI later when we upgrade, assuming we can live with this event-crawling ;)

Let me know if you're totally against this, want more documentation, and/or can live w/ this until we upgrade Paragon :)

**Review by stvstnfrd — COMMENTED** [review]
@davidjoy Let me know if you can live w/ my "close the dropdown" hack.

I think I'd like to punt on the naming nits for now, in part to avoid last-minute churn that could trigger breakage, but also to spare both of us from a naming conversation now :D 

I'm going to take the "out" that this isn't a full standard (yet?) in either the React community or this codebase, but I'm a big fan of predictable naming conventions, so happy to talk about adopting some convention(s) moving forward.

File: src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx (hunk 37 above)
Comment (davidjoy) [reply src/instructor-toolbar/masquerade-widget/MasqueradeWidgetOption.jsx:16]: Shipppppit!

**Review by davidjoy — APPROVED** [review]
